cmd/gardener-scheduler/app/gardener_scheduler.go

```diff
@@ -0,0 +1,304 @@
+		return nil, err
+	}
+
+	// Set up leader election if enabled and prepare event recorder.
+	var (
+		leaderElectionConfig *leaderelection.LeaderElectionConfig
+		recorder             = utils.CreateRecorder(k8sGardenClient.Kubernetes(), "gardener-scheduler")
+	)
+	if cfg.LeaderElection.LeaderElect {
+		leaderElectionConfig, err = utils.MakeLeaderElectionConfig(cfg.LeaderElection.LeaderElectionConfiguration, cfg.LeaderElection.LockObjectNamespace, cfg.LeaderElection.LockObjectName, k8sGardenClientLeaderElection, recorder)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &GardenerScheduler{
+		Config:                 cfg,
+		Logger:                 logger,
+		Recorder:               recorder,
+		K8sGardenClient:        k8sGardenClient,
+		K8sGardenInformers:     gardeninformers.NewSharedInformerFactory(k8sGardenClient.Garden(), 0),
+		K8sGardenCoreInformers: gardencoreinformers.NewSharedInformerFactory(k8sGardenClient.GardenCore(), 0),
+		KubeInformerFactory:    kubeinformers.NewSharedInformerFactory(k8sGardenClient.Kubernetes(), 0),
+		LeaderElection:         leaderElectionConfig,
+	}, nil
+}
+
+func (g *GardenerScheduler) cleanup() {
+	if err := os.RemoveAll(schedulerconfigv1alpha1.DefaultDiscoveryDir); err != nil {
+		g.Logger.Errorf("Could not cleanup base discovery cache directory: %v", err)
+	}
+}
+
+// Run runs the Gardener Scheduler. This should never exit.
+func (g *GardenerScheduler) Run(ctx context.Context) error {
+	defer g.cleanup()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Prepare a reusable run function.
+	run := func(ctx context.Context) {
+		g.startScheduler(ctx)
+	}
+
+	// Start HTTP server (HTTPS not needed because no webhook server is needed at the moment)
+	go server.ServeHTTP(ctx, g.Config.Server.HTTP.Port, g.Config.Server.HTTP.BindAddress)
+	handlers.UpdateHealth(true)
+
+	// If leader election is enabled, run via LeaderElector until done and exit.
+	if g.LeaderElection != nil {
+		g.LeaderElection.Callbacks = leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(_ context.Context) {
+				g.Logger.Info("Acquired leadership, starting scheduler.")
+				run(ctx)
+			},
+			OnStoppedLeading: func() {
+				g.Logger.Info("Lost leadership, terminating.")
+				cancel()
+			},
+		}
+
+		leaderElector, err := leaderelection.NewLeaderElector(*g.LeaderElection)
+		if err != nil {
+			return fmt.Errorf("couldn't create leader elector: %v", err)
+		}
+
+		leaderElector.Run(ctx)
+		return nil
+	}
+	run(ctx)
+	return nil
+}
+
+func (g *GardenerScheduler) startScheduler(ctx context.Context) {
+	gardenerScheduler := controller.NewGardenerScheduler(g.K8sGardenClient, g.K8sGardenInformers, g.Config, g.Recorder)
+
+	// Initialize the Controller metrics collection.
+	gardenmetrics.RegisterControllerMetrics(gardenerScheduler)
+
+	// scheduling shoots is not a very frequent operation (compared to scheduling pods), so 1 worker should be fine
+	go gardenerScheduler.Run(ctx, g.K8sGardenInformers, 1)
```

Comment:
Sorry for seeing this that late: I guess 1 worker might not be sufficient (just imagine multiple shoots being requested at once) - In general, this should be
* configurable
* by default sth like 5